crates/core/src/address.rs

```diff
@@ -410,17 +410,25 @@ impl TryFrom<&Signer> for Address {
     type Error = DecodeError;
 
     fn try_from(signer: &Signer) -> Result<Self> {
-        // The given address should be an address or payment address. When
-        // sending a token from a spending key, it has been already
-        // replaced with the MASP address.
-        Address::decode(signer.as_ref()).or(
-            match crate::masp::PaymentAddress::from_str(signer.as_ref()) {
-                Ok(_) => Ok(MASP),
-                Err(_) => Err(DecodeError::InvalidInnerEncoding(format!(
-                    "Invalid address for IBC transfer: {signer}"
-                ))),
-            },
-        )
+        let signer: &str = signer.as_ref();
+
+        #[cold]
+        fn is_znam(signer: &str) -> bool {
+            crate::masp::PaymentAddress::from_str(signer).is_ok()
+        }
+
+        Address::decode(signer).map_err(|err| {
```

Comment:
Should we also adjust/check the `receiver` argument in the sdk that eventually ends up in the `PacketData`? It seems to me these trait implementation is used in the vp but by the time this check runs we have already leaked information

Reply:
hm, the receiver of the ibc transfer is a namada shielded account. I think the best we can do is validate the tx before it enters the mempool, which might be a bit overkill. we can't validate the receiver in the packet sent to namada. it's gonna be set from the ui/cli of whatever chain sends the packet.